docs/_docs/02_features/keep-your-remote-state-configuration-dry.md

```diff
@@ -135,6 +137,58 @@ The child `.hcl` file’s `terraform` settings will be merged into the parent fi
 
 Other settings in the child `.hcl` file override the respective settings in the parent.
 
+
+### Using the generate property to generate terraform code for managing remote state
+
+While the default way terragrunt manages remote state is through `terraform init` with `-backend-config`, you can also
+use the `generate` property to configure terragrunt to geneerate a `.tf` file in the terraform working directory with the
+backend configuration.
+
+The `generate` property is an object that accepts two parameters:
+
+- `path`: The path where the generated file should be written. If a relative path, it'll be relative to the Terragrunt
+  working dir (where the terraform code lives).
+- `if_exists`: What to do if a file already exists at `path`. Valid values are: `overwrite` (overwrite the existing
+  file), `skip` (skip code generation and leave the existing file as-is), `error` (exit with an error).
+
+For example, here is a version of the root `remote_state` configuration with the `generate` property:
+
+```hcl
+remote_state {
+  backend = "s3"
+  generate = {
+    path      = "backend.tf"
+    if_exists = "overwrite"
```

Comment:
I just had one realizationwith the `generate` helper, both in `remote_state` and in general, in terms of how the `if_exists` logic works. Let's say you are generating a `backend.tf` into some `.terragrunt-cache/xxx` folder. There are three ways a `backend.tf` file could already exist in that folder:

1. It's part of the code you downloaded from the `source` URL.
1. It's one of the files in the working directory where you ran `terragrunt apply`, and Terragrunt copied it from that folder into the `.terragrunt-cache/xxx` folder.
1. You generated `backend.tf` on a previous run.

Some important questions about these three cases:

1. Does code generation happen before or after case (1)?
1. Does code generation happen before or after case (2)?
1. Should `if_exists` provide a way to differentiate between case (3), which will happen all the time and is normal, versus cases (1) and (2) which are more rare and probably indicative of a problem?

Reply:
Code generation happens after case (2). There is a test case that checks this behavior by having a `backend.tf` file in the terragrunt working dir and setting a `generate` block having `if_exists = "skip"`. The test then makes sure it doesn't overwrite the `backend.tf` file. Similarly, there is a case that has a similar set up with `if_exists = "overwrite"` to make sure it overwrites the `backend.tf`.

> Should if_exists provide a way to differentiate between case (3), which will happen all the time and is normal, versus cases (1) and (2) which are more rare and probably indicative of a problem?

This is an interesting question. I think it should, but I can't come up with a clean way to implement this. The only way where I can think of a stable way to implement this is to have a comment at the top of the file when Terragrunt generates the code that can be used as a signature. If the file contains that, then we know it is a terragrunt generated file.

Let me implement that for now.

Reply:
Done: a6e8f80

Reply:
Ah, nice work with the comment. That's useful too so that users can easily tell it's generated code.